core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -1091,14 +1091,11 @@ object IO extends Serializable {
   )(release: (A, ExitResult[E, B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
     Ref[Option[(A, Fiber[E, B])]](None).flatMap { m =>
       (for {
-        f <- acquire.flatMap(a => use(a).fork.flatMap(f => m.set(Some(a -> f)).const(f))).uninterruptibly
+        f <- acquire.flatMap(a => use(a).fork.peek(f => m.set(Some(a -> f)))).uninterruptibly
         b <- f.join
       } yield b).ensuring(m.get.flatMap(_.fold(IO.unit) {
         case (a, f) =>
-          f.poll.flatMap {
-            case Some(r) => release(a, r)
-            case None    => f.interrupt *> f.observe.flatMap(release(a, _))
-          }
+          f.interrupt *> f.observe.flatMap(release(a, _))
```

Comment:
How about:
```scala
trait Fiber[E, A] {
  def interrupt: IO[Nothing, ExitResult[E, A]]
}
```
?

Reply:
Also, this will slow down the happy path, but not by much, right?

Reply:
This is a wonderful idea! Makes `interrupt` more useful and shows you — in the types — that it waits until the interruption has completed.

This change will make things much faster. `interrupt` returns instantaneously if the fiber is already done.